components/brave_wallet/browser/tx_state_manager.cc

```diff
@@ -107,41 +108,45 @@ bool TxStateManager::ValueToTxMeta(const base::Value::Dict& value,
   return true;
 }
 
-TxStateManager::TxStateManager(PrefService* prefs)
-    : prefs_(prefs), weak_factory_(this) {}
+TxStateManager::TxStateManager(PrefService* prefs, TxStorageDelegate* delegate)
+    : prefs_(prefs), delegate_(delegate), weak_factory_(this) {
+  DCHECK(delegate);
+}
 
 TxStateManager::~TxStateManager() = default;
 
-void TxStateManager::AddOrUpdateTx(const TxMeta& meta) {
-  ScopedDictPrefUpdate update(prefs_, kBraveWalletTransactions);
-  base::Value::Dict& dict = update.Get();
+bool TxStateManager::AddOrUpdateTx(const TxMeta& meta) {
+  if (!delegate_->IsInitialized()) {
+    return false;
+  }
   const std::string path =
       base::JoinString({GetTxPrefPathPrefix(meta.chain_id()), meta.id()}, ".");
-
-  bool is_add = dict.FindByDottedPath(path) == nullptr;
-  dict.SetByDottedPath(path, meta.ToValue());
+  auto& txs = delegate_->GetTxs();
+  bool is_add = txs.FindByDottedPath(path) == nullptr;
+  txs.SetByDottedPath(path, meta.ToValue());
+  delegate_->ScheduleWrite();
   if (!is_add) {
     for (auto& observer : observers_) {
       observer.OnTransactionStatusChanged(meta.ToTransactionInfo());
     }
-    return;
-  }
+  } else {
+    for (auto& observer : observers_) {
+      observer.OnNewUnapprovedTx(meta.ToTransactionInfo());
+    }
 
-  for (auto& observer : observers_) {
-    observer.OnNewUnapprovedTx(meta.ToTransactionInfo());
+    // We only keep most recent 1k confirmed plus rejected tx metas per network
+    RetireTxByStatus(meta.chain_id(), mojom::TransactionStatus::Confirmed,
+                     kMaxConfirmedTxNum);
+    RetireTxByStatus(meta.chain_id(), mojom::TransactionStatus::Rejected,
+                     kMaxRejectedTxNum);
   }
-
-  // We only keep most recent 10 confirmed and rejected tx metas per network
-  RetireTxByStatus(meta.chain_id(), mojom::TransactionStatus::Confirmed,
-                   kMaxConfirmedTxNum);
-  RetireTxByStatus(meta.chain_id(), mojom::TransactionStatus::Rejected,
-                   kMaxRejectedTxNum);
+  return true;
 }
 
 std::unique_ptr<TxMeta> TxStateManager::GetTx(const std::string& chain_id,
                                               const std::string& id) {
-  const auto& dict = prefs_->GetDict(kBraveWalletTransactions);
-  const base::Value::Dict* value = dict.FindDictByDottedPath(
+  const auto& txs = delegate_->GetTxs();
```

Comment:
For consistency with other methods accessing `delegate_->GetTxs()`
```
if (!delegate_->IsInitialized()) {
  return nullptr;
}
```

Reply:
fixed in 330c37ab4a9a518d77a58769424ffe8b45e84776